runtime/include/chpl-mem.h

```diff
@@ -192,4 +178,20 @@ void chpl_mem_layerFree(void*, int32_t lineno, int32_t filename);
 
 #endif // LAUNCHER
 
+// Provide handles to instrument Chapel calls to memcpy and memmove. For memcpy
+// we've named it to explicit indicate overlap isn't permitted since we've
+// historically been pretty careless about that.
+static inline
+void* chpl_no_overlap_memcpy(void* dest, const void* src, size_t num)
+{
+  assert(dest != src || num == 0);
```

Comment:
I was originally with you on adding this check, but now I'm not positive this is legal. I think relational operators between pointers is only defined if they point to the same object and is UB otherwise.

@dmk42 mind weighing in?

Reply:
That's why I cast everything to numeric types.

Reply:
This is a really tricky area.  It turns out that casting to integers makes it worse.  Once two pointers are cast to integers, there is no guarantee that any relationship between the integers has anything to do with the relationship between the pointers, except that `==` must work.

The only thing that is guaranteed to work is too slow.  (Theoretically it could be optimized, but I don't know of any compilers that do.)
```c
char *d = dest;
char *s = src;
bool overlap = false;
for (size_t i = 0; i < num; ++i) {
  if (d + i == s || s + i == d) {
    overlap = true;
    break;
  }
}
```

The only other way to write the comparison in C is to use nonportable extensions.  The only nonportable extension that helps in this case and is universally available (accidentally portable) is to turn off optimization.  The overlap check could be done in an external function, and the external function could live in a source file that is compiled with optimization turned off.

Even then, the solution will only be portable to systems with a flat address space.  If Chapel is ported to a non-flat-address-space machine, a strategy specific to that machine will probably be needed.

It seems to me that it would be better to avoid the check altogether, and just use `memmove`.

Reply:
@dmk42 - isn't it the case that the code I suggested would just miss cases? Would the assert possibly fire when it shouldn't? Wouldn't in work in practice on the systems we test on?

Reply:
It could have both false positives and false negatives, because of optimization.  Basically, it can't be depended on.

Reply:
I find this very surprising, to say the least. I'd be interested in arguing more about it in the future. But that won't help Elliot near-term.

For this PR, we could:
 * consider enabling the check on our test systems (in a limited setting)
 * not add the proposed improvement to the check

Reply:
It depends on more than just the hardware.  Each version of each target compiler is allowed to make different choices because of optimization.

I guess I don't understand the purpose of the check.  If there might be an overlap, why is `memmove` not used?

Reply:
This PR changes a bunch of calls to use `memmove` but it'd be nice if we could have more confidence that we don't accidentally call `memcpy` when they do overlap. So it'd be nice if we could verify that in nightly testing, somehow.

Reply:
OK, that makes sense.  Since this is for testing, I'd recommend putting the test in an external function that is compiled without optimization.  It will only work for flat address spaces, but that's all we test on right now.

Reply:
@ronawho - Up above, @mppf notes that not adding the proposed improvement to the check would be an option.  I'm OK with that if you think that just having the `==` comparison would catch the most likely cases.